x-pack/plugins/transform/public/app/hooks/use_index_data.ts

```diff
@@ -86,73 +77,75 @@ export const useIndexData = (
-      let populatedDataViewFields = populatedFields ? [...populatedFields] : [];
-      let isMissingFields = populatedDataViewFields.length === 0;
-
-      // If populatedFields are not provided, make own request to calculate
-      if (populatedFields === undefined) {
-        setErrorMessage('');
-        setStatus(INDEX_STATUS.LOADING);
-
-        const esSearchRequest = {
-          index: indexPattern,
-          body: {
-            fields: ['*'],
-            _source: false,
-            query: {
-              function_score: {
-                query: defaultQuery,
-                random_score: {},
-              },
-            },
-            size: 500,
+  // Fetch 500 random documents to determine populated fields.
+  // This is a workaround to avoid passing potentially thousands of unpopulated fields
+  // (for example, as part of filebeat/metricbeat/ECS based indices)
+  // to the data grid component which would significantly slow down the page.
+  const {
+    error: dataViewFieldsError,
+    data: dataViewFieldsData,
+    isError: dataViewFieldsIsError,
+    isLoading: dataViewFieldsIsLoading,
+  } = useDataSearch(
+    {
+      index: indexPattern,
+      body: {
+        fields: ['*'],
+        _source: false,
+        query: {
+          function_score: {
+            query: defaultQuery,
+            random_score: {},
           },
-        };
+        },
+        size: 500,
+      },
+    },
+    // Check whether fetching should be enabled
+    // If populatedFields are not provided, make own request to calculate
+    !!populatedFields && !(dataView.timeFieldName !== undefined && timeRangeMs === undefined)
+  );
 
-        const resp = await dataSearch(esSearchRequest, abortController.signal);
+  useEffect(() => {
+    if (dataViewFieldsIsLoading && !dataViewFieldsIsError) {
+      setErrorMessage('');
+      setStatus(INDEX_STATUS.LOADING);
+    } else if (dataViewFieldsError !== null) {
+      setErrorMessage(getErrorMessage(dataViewFieldsError));
+      setStatus(INDEX_STATUS.ERROR);
+    } else if (
+      !dataViewFieldsIsLoading &&
+      !dataViewFieldsIsError &&
+      dataViewFieldsData !== undefined
+    ) {
+      const isCrossClusterSearch = indexPattern.includes(':');
+      const isMissingFields = dataViewFieldsData.hits.hits.every(
+        (d) => typeof d.fields === 'undefined'
+      );
 
-        if (!isEsSearchResponse(resp)) {
-          setErrorMessage(getErrorMessage(resp));
-          setStatus(INDEX_STATUS.ERROR);
-          return;
-        }
-        const docs = resp.hits.hits.map((d) => getProcessedFields(d.fields ?? {}));
-        isMissingFields = resp.hits.hits.every((d) => typeof d.fields === 'undefined');
+      setCcsWarning(isCrossClusterSearch && isMissingFields);
+      setStatus(INDEX_STATUS.LOADED);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [dataViewFieldsData, dataViewFieldsError, dataViewFieldsIsError, dataViewFieldsIsLoading]);
```

Comment:
The entire effect looks very complex. Essentially it derives error message and status.
Can we get rid of the state and use memo instead? e.g.

const errorMessage = useMemo(() => {...}, [dataViewFieldsIsLoading, dataViewFieldsIsError]);